fix subclass custom sizes dynamic shapes caching

bdhirsh opened the pull request:
This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with @ezyang and @soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #109023
* #109632
* #109662
* __->__ #108654

bdhirsh committed:
[not ready yet] fix subclass custom sizes dynamic shapes caching

[ghstack-poisoned]

bdhirsh committed:
Update on "[not ready yet] fix subclass custom sizes dynamic shapes caching"




[ghstack-poisoned]

bdhirsh committed:
Update on "[not ready yet] fix subclass custom sizes dynamic shapes caching"




[ghstack-poisoned]

bdhirsh committed:
Update on "[not ready yet] fix subclass custom sizes dynamic shapes caching"




[ghstack-poisoned]

bdhirsh committed:
Update on "[not ready yet] fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh commented:
@ezyang this is ready for review now (I got the tests in AOTAutograd that I was running locally to finally pass - although I'm waiting on the latest round of CI to confirm that python functionalization tests are passing in all cases)

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -555,6 +555,150 @@ static void set_tensor_attr_with_capsule(
+      PyCapsule_GetPointer(curr_buffer_pyobj.ptr(), nullptr);
+  auto curr_buffer = reinterpret_cast<T*>(buffer_pycapsule);
+  return c10::ArrayRef<T>(curr_buffer, new_size);
+}
```
Meh. Here's a better idea: literally store a SizesAndStrides object inside the capsule, and then reuse the implementation there. Just need some placement-new/delete to get things setup initially.

bdhirsh replied:
One reason that seems a bit painful is that `SizesAndStrides` stores the sizes and strides together in a single location in memory. The lazy updating of the buffer during metadata mutations makes that kind of painful here, because the flow in the PR as written is currently:

```
# For a normal tensor this will eagerly cause the SizesAndStrides buffer to do a reallocation
# But for subclasses, we have an independent size and stride buffer that will be lazily reallocated on first access
x.resize_(...)
...
# For subclasses, at this point, we re-allocate the size buffer when we detect that our buffer is the wrong size
sizes = x.size()

# For subclasses, at this point, we re-allocate the stride buffer when we detect that our buffer is the wrong size
sizes = x.size()
```

bdhirsh replied:
But using a SizeAndStride object here seems to imply that we should actually store the sizes and strides together into that single object. That would mean that when a user calls `sym_size()` and we query the subclass for the updated size, we would also need to query for the updated stride, so we can stuff them both into the `SizeAndStride` object that we're using.

Is that what you were thinking? Or were you more just thinking of only re-using `SizeAndStrides` for the SmallVector optimization, and not to actually store the two of them together.

ezyang replied:
Hmm. Yes, I think I am concretely proposing you force both a size and stride query whenever one query occurs.

bdhirsh replied:
ok now I am tempted to try to root-cause the bug that required me to add the SmallVector optimization :) Failing that, I'll do the SizeAndStride update.

(Actually, double checking - if we can avoid needing to make the SmallVector optimization happen here, are you ok with a vanilla `new T[updated_size]` call?)

ezyang replied:
Yeah, if you don't need to exactly mimic SizeAndStrides, just something simple is good. But if you DO want to mimic SizeAndStrides, might as well do the real thing.

By the way, you could avoid calling the stride query if you don't resize.

bdhirsh replied:
aaand... tests still pass when I remove the small vector optimization. I thought I checked this before I added it, but I must have also fixed another bug at the same time or something...

PR updated to not do the SmallVector thing and just do `new T[updated_size]`.

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

soulitzer commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -770,16 +894,10 @@ c10::SymIntArrayRef ConcretePyInterpreterVTable::sym_strides(
-  for (auto it = out.begin(); it != out.end(); ++it, ++idx) {
-    ptr[idx] = py::cast<c10::SymInt>(*it);
-  }
-  set_tensor_attr_with_capsule(self, capsule, "_sym_strides_capsule");
```
Don't forget to remove this func :)

bdhirsh replied:
I'm actually still using this func (we still need to set a pycapsule buffer on the tensor)

soulitzer replied:
Ah whoops!

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh commented:
Ok, re-ran with my SmallVector optimization turned off and I see ASAN complaining here, in the `_batch_norm_impl_index` composite kernel (pasted below).

I'm a bit confused by the output, because what I'd expect is that we must be storing a `SymIntArrayRef` somewhere, performing a resize (of <= 5 elements), and then trying to use the `SymIntArrayRef` that we stashed somewhere. That doesn't seem to be the case in that kernel though: it takes in no size-y inputs, and just grabs symints directly (e.g. `auto num_features = input.sym_sizes()[1]`).

For now... I'm tempted to add back my ugly `SmallVector` optimization, and *promise* to do the real cleanup of re-writing it to do placement new `SizesAndStrides` later (per offline discussion w @ezyang yesterday, lmk if that's still ok with you)

https://pipelinesghubeus21.actions.githubusercontent.com/mBh68xKhi8LyM7tp3vECvYXNFvuV4gyVGgmYCteuEZP9JH92QN/_apis/pipelines/1/runs/3373307/signedlogcontent/790?urlExpires=2023-09-15T21%3A13%3A51.4327798Z&urlSigningMethod=HMACV1&urlSignature=tDeX7ZqaARVU5NNwyr5yYqqkWq3A2j4z8FFdqYwGr0Q%3D

```
2023-09-15T18:18:00.9370118Z ==623==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020001d95c0 at pc 0x7f835976de2a bp 0x7ffe2acdedb0 sp 0x7ffe2acdeda8
2023-09-15T18:18:00.9370226Z READ of size 8 at 0x6020001d95c0 thread T0
2023-09-15T18:18:00.9370589Z     #0 0x7f835976de29 in c10::SymInt::SymInt(c10::SymInt const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x8e56e29)
2023-09-15T18:18:00.9371278Z     #1 0x7f835a5627b3 in at::native::_batch_norm_impl_index(at::Tensor const&, c10::optional<at::Tensor> const&, c10::optional<at::Tensor> const&, c10::optional<at::Tensor> const&, c10::optional<at::Tensor> const&, bool, double, double, bool) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x9c4b7b3)
...
...
2023-09-15T18:18:00.9507158Z 0x6020001d95c0 is located 0 bytes to the right of 16-byte region [0x6020001d95b0,0x6020001d95c0)
2023-09-15T18:18:00.9507256Z allocated by thread T0 here:
2023-09-15T18:18:00.9507570Z     #0 0x7f8381edf9ed in operator new[](unsigned long) (/usr/lib/llvm-12/lib/clang/12.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xcd9ed)
2023-09-15T18:18:00.9508159Z     #1 0x7f837972911e in c10::ArrayRef<c10::SymInt> (anonymous namespace)::get_set_cached_attr<c10::SymInt>(c10::TensorImpl const*, char const*, pybind11::object const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x259a11e)
2023-09-15T18:18:00.9508669Z     #2 0x7f837971d4b8 in (anonymous namespace)::ConcretePyInterpreterVTable::sym_sizes(c10::TensorImpl const*) const (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x258e4b8)
2023-09-15T18:18:00.9509100Z     #3 0x7f83799391d8 in torch::autograd::InputMetadata::compute_variant_shape(at::Tensor const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x27aa1d8)
2023-09-15T18:18:00.9509516Z     #4 0x7f8379938dd9 in torch::autograd::InputMetadata::InputMetadata(at::Tensor const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x27a9dd9)
2023-09-15T18:18:00.9510115Z     #5 0x7f8379938903 in torch::autograd::InputMetadata& c10::SmallVectorImpl<torch::autograd::InputMetadata>::emplace_back<at::Tensor const&>(at::Tensor const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x27a9903)
2023-09-15T18:18:00.9510509Z     #6 0x7f837993879b in torch::autograd::Node::add_input_metadata(at::Tensor const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x27a979b)
2023-09-15T18:18:00.9510949Z     #7 0x7f8379937f27 in torch::autograd::set_history(at::Tensor&, std::shared_ptr<torch::autograd::Node> const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x27a8f27)
2023-09-15T18:18:00.9511422Z     #8 0x7f837992cab1 in torch::autograd::THPVariable__mirror_autograd_meta_to(_object*, _object*, _object*) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x279dab1)
2023-09-15T18:18:00.9511681Z     #9 0x507386 in cfunction_call /usr/local/src/conda/python-3.9.18/Objects/methodobject.c:543:19
2023-09-15T18:18:00.9511690Z 
2023-09-15T18:18:00.9512210Z SUMMARY: AddressSanitizer: heap-buffer-overflow (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x8e56e29) in c10::SymInt::SymInt(c10::SymInt const&)
```

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+  size_t curr_size = 0;
+  if (PyObject_HasAttrString(mb_obj.value(), buffer_len_attr_name.c_str())) {
+    auto buffer_pyobj =
+        py::handle(mb_obj.value()).attr(buffer_len_attr_name.c_str());
```
You keep referencing mb_obj.value(), put it in a variable

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+    TORCH_INTERNAL_ASSERT(PyObject_HasAttrString(
+        mb_obj.value(), buffer_issmallvector_attr_name.c_str()));
+    (PyObject_HasAttrString(
+        mb_obj.value(), buffer_issmallvector_attr_name.c_str()));
```
What's this

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+        mb_obj.value(), buffer_issmallvector_attr_name.c_str()));
+    auto issmallvector_pyobj =
+        py::handle(mb_obj.value()).attr(buffer_issmallvector_attr_name.c_str());
+    is_smallvector = py::cast<bool>(issmallvector_pyobj);
```
I'm not sure why you need this extra attribute, can't you infer if it's small or not just by looking at the size?

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+  // SmallVector starts out by allocating a size-5 buffer. Any resizes on the
+  // tensor that are <= 5 elements will not reallocate the buffer. But once we
+  // hit a size that is > 5 elements, we will re-allocate on every future
+  // resize. NOTE: Ideally, we shouldn't need to model the SmallVector
```
This doesn't seem quite accurate. If you go back down to <= 5 elements, you'll go back into the small vector regime. There isn't any hysteresis (every future resize)

bdhirsh replied:
Ah whoops, I didn't read the SmallVector code carefully enough. Thanks, will update.

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+    // (due to a metadata mutation).
+    // This is technically not thread safe (maybe we should put the GIL here?),
+    // but *only* if we actually have to update, which only happens if there's a
+    // metadata mutation.
```
This is not quite accurate because the update here can be delayed. So it could be that the original metadata mutation was thread safe, but because we lazily update, we don't end up mutating until we're out of the critical region and there is possibility for data race.

bdhirsh replied:
to confirm: do you think this is something we should try to fix?

Agreed that the comment is inaccurate, I'll fix (you're pointing out that it could be the case that a metadata mutation happened in eager mode in a thread safe way, but our lazy update ends up not being thread safe)

ezyang replied:
It's kind of hard to fix, because you have to do the update lazily. Having a lock is probably the only way to do it. Comment update is OK for this PR.

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+             curr_buffer[idx] != actual_val) ||
+            (curr_buffer[idx].is_heap_allocated() &&
+             curr_buffer[idx].toSymNodeImplUnowned() !=
+                 actual_val.toSymNodeImplUnowned())) {
```
Nyeh, we should probably just add a method to SymInt that does this, you literally just want to look and see if the bit pattern is the same, you really shouldn't need a conjunction of tests

bdhirsh replied:
Definitely fair, will add

bdhirsh replied:
Bikeshedding the name: `SymInt::identity_equals`?

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+    // (sized to 5)
+    py::handle(mb_obj.value()).attr(buffer_issmallvector_attr_name.c_str()) =
+        set_to_smallvector;
+  } else {
```
I'd feel better with an assert that cur size says it is large enough to hold all the writes

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+      ((!is_smallvector || new_size > 5) && curr_size != new_size);
+  if (needs_resize) {
+    auto set_to_smallvector = false;
+    // If our current buffer is not the right size (eithr because we haven't
```
either

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+    // will trash the previous buffer if there already was one, invalidating any
+    // existing SymIntArrayRef's from an old .sym_size() call.
+    auto new_buffer_size = new_size;
+    if (!is_buffer_allocated && new_size <= 5) {
```
yeah so you have to always do this, even if the buffer is allocated

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh commented:
PR is updated

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -562,7 +562,133 @@ static void set_tensor_attr_with_capsule(
+  bool is_buffer_allocated = false;
+  size_t curr_size = 0;
+  if (PyObject_HasAttrString(tensor_obj, buffer_len_attr_name.c_str())) {
+    auto buffer_pyobj =
```
nit: it's weird to call this `buffer_pyobj` since it's not actually the buffer lol

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix subclass custom sizes dynamic shapes caching"

This PR fixes the ownership/lifetime handling for tensor subclasses that override sizes/strides, when tensors get resized.

This is needed now, because `FunctionalTensor` is a subclass that has a custom size/stride (so it can plumb requests to its inner tensor), and is also a core piece of infra (it's used during tracing in AOTAutograd, which means that metadata mutation and resizing that happens to work with torch.compile today needs to work with FunctionalTensor).

After a bunch of discussion with ezyang and soulitzer, I updated `PyInterpreter::sym_sizes()` (and friends) so that:
(1) They allocate a py::capsule buffer and stash it on the tensor on the first call to size/stride
(2) On a size/stride call where we noticed that the number of **dimensions** on the tensor has changed (so our buffer it stale), we re-allocate the buffer
(3) On a size/strude cal where we notice that the number of dimensions is the same, but the values are different (this happens whenever a tensor experiences a metadata mutation, like `.transpose_()`), we inplace-modify the buffer and put the new ints/symints into it

I also ended up doing the SmallVector optimization, which was required to fix some tests in AOTAutograd. Ideally we should look into those tests, and nail down the parts of our codebase that rely on SmallVector not re-allocating on a resize... but I'm saving this for a followup.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
